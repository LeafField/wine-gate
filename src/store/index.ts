import { create } from "zustand";

type User = {
  username: string;
  id: string;
  email: string | undefined;
};

type Menu = "close" | "member" | "free" | "category";

type Store = {
  user: User | null;
  setUser: (user: User) => void;
  logout: () => void;
  menu: Menu;
  setMenu: (menu: Menu) => void;
};

export const useStore = create<Store>((set) => ({
  user: null,
  setUser: (user) => set({ user }),
  logout: () => set({ user: null }),
  menu: "close",
  setMenu: (menu) => set({ menu }),
}));
