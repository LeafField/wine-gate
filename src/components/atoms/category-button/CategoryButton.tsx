"use client";
import React, { FC } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faList } from "@fortawesome/free-solid-svg-icons";
import { useStore } from "../../../store";
import CloseButton from "../close-button/CloseButton";

const CategoryButton: FC = () => {
  const { menu, setMenu } = useStore();

  const handleClick = () => {
    setMenu("category");
  };

  if (menu === "category") {
    return <CloseButton rightBorder={false} />;
  } else {
    return (
      <button
        type="button"
        className="flex w-full flex-col items-center justify-center"
        onClick={handleClick}
      >
        <FontAwesomeIcon
          className="text-menu-icon text-white"
          icon={faList}
          aria-hidden={"true"}
        />
        <span className="text-menu-title text-white">カテゴリー検索</span>
      </button>
    );
  }
};
// 0.75rem = 12px
export default CategoryButton;
