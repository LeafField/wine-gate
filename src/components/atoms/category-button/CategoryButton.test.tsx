import {
  screen,
  render,
  waitFor,
  renderHook,
  act,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import CategoryButton from "./CategoryButton";
import { useStore } from "../../../store";

describe("CategoryButtonの結合テスト", () => {
  beforeAll(() => {
    const { result } = renderHook(() => useStore());
    act(() => {
      result.current.setMenu("close");
    });
  });

  test("初期状態ではカテゴリー検索ボタンが表示されている", () => {
    const { container } = render(<CategoryButton />);
    expect(screen.getByRole("button")).toHaveTextContent("カテゴリー検索");
    expect(container).toMatchSnapshot();
  });

  test("カテゴリー検索ボタンをクリックすると、閉じるボタンに変わる", async () => {
    render(<CategoryButton />);
    await userEvent.click(screen.getByRole("button"));
    return expect(screen.getByRole("button")).toHaveTextContent("閉じる");
  });

  test("閉じるボタンをクリックすると、カテゴリー検索ボタンに変わる", async () => {
    render(<CategoryButton />);
    await userEvent.click(screen.getByRole("button", { name: "閉じる" }));
    await waitFor(() => {
      expect(screen.getByRole("button")).toHaveTextContent("カテゴリー検索");
    });
  });
});
