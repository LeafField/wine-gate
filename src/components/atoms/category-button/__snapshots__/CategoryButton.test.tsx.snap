// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryButtonの結合テスト 初期状態ではカテゴリー検索ボタンが表示されている 1`] = `
<div>
  <button
    aria-controls="category-panel"
    aria-expanded="false"
    class="flex w-full flex-col items-center justify-center"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-list text-menu-icon text-white"
      data-icon="list"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M40 48C26.7 48 16 58.7 16 72l0 48c0 13.3 10.7 24 24 24l48 0c13.3 0 24-10.7 24-24l0-48c0-13.3-10.7-24-24-24L40 48zM192 64c-17.7 0-32 14.3-32 32s14.3 32 32 32l288 0c17.7 0 32-14.3 32-32s-14.3-32-32-32L192 64zm0 160c-17.7 0-32 14.3-32 32s14.3 32 32 32l288 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-288 0zm0 160c-17.7 0-32 14.3-32 32s14.3 32 32 32l288 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-288 0zM16 232l0 48c0 13.3 10.7 24 24 24l48 0c13.3 0 24-10.7 24-24l0-48c0-13.3-10.7-24-24-24l-48 0c-13.3 0-24 10.7-24 24zM40 368c-13.3 0-24 10.7-24 24l0 48c0 13.3 10.7 24 24 24l48 0c13.3 0 24-10.7 24-24l0-48c0-13.3-10.7-24-24-24l-48 0z"
        fill="currentColor"
      />
    </svg>
    <span
      class="text-menu-title text-white"
    >
      カテゴリー検索
    </span>
  </button>
</div>
`;
